### PR TITLE
Improve tests for parsing x509 SubjectKeyId and AuthorityKeyId

### PR DESCRIPTION
**Pull request by mprse**
## Description

Resolves: https://github.com/Mbed-TLS/mbedtls/issues/6985

This PR is build on top of: PR https://github.com/Mbed-TLS/mbedtls/pull/6866. First commit of this PR: d2507ff5d1ae3e8b05e0b8345e8497d0d78d33ed.

Provides make file generated certificates for tests and improves test by checking values of tested cert fields (not length only).

## Gatekeeper checklist

- [X] **changelog** Not needed, test improvements only
- [X] **backport** not required
- [X] **tests** provided

## Notes for the submitter

Please refer to the [contributing guidelines](https://github.com/Mbed-TLS/mbedtls/blob/development/CONTRIBUTING.md), especially the
checklist for PR contributors.

### CODE REVIEW

**Comment by mprse** [comment]
Coverage report below:

image

image

**Comment by mprse** [comment]
@DemiMarie @AndrzejKurek This one is ready for review.

File: tests/suites/test_suite_x509parse.data [hunk 1]
```diff
@@ -3280,37 +3280,61 @@ depends_on:MBEDTLS_X509_CRT_PARSE_C
...
+depends_on:MBEDTLS_MD_CAN_SHA256:MBEDTLS_RSA_C
+x509_crt_parse_authoritykeyid:"data_files/authorityKeyId_subjectKeyId_issuer_tag2_malformed.crt.der":"":"":"":MBEDTLS_ERR_X509_INVALID_EXTENSIONS+MBEDTLS_ERR_ASN1_UNEXPECTED_TAG
+
+X509 CRT parse Authority Key Id - Wrong Serial Number Tag
+depends_on:MBEDTLS_MD_CAN_SHA256:MBEDTLS_RSA_C
+x509_crt_parse_authoritykeyid:"data_files/authorityKeyId_subjectKeyId_sn_tag_malformed.crt.der":"":"":"":MBEDTLS_ERR_X509_INVALID_EXTENSIONS+MBEDTLS_ERR_ASN1_UNEXPECTED_TAG
+
+X509 CRT parse Authority Key Id - Wrong Serial Number Tag length
+depends_on:MBEDTLS_MD_CAN_SHA256:MBEDTLS_RSA_C
+x509_crt_parse_authoritykeyid:"data_files/authorityKeyId_subjectKeyId_sn_len_malformed.crt.der":"":"":"":MBEDTLS_ERR_X509_INVALID_EXTENSIONS+MBEDTLS_ERR_ASN1_LENGTH_MISMATCH
```
Comment (DemiMarie) [inline tests/suites/test_suite_x509parse.data:3340]:
```suggestion
x509_crt_parse_authoritykeyid:"data_files/authorityKeyId_subjectKeyId_sn_len_malformed.crt.der":"":"":"":MBEDTLS_ERR_X509_INVALID_EXTENSIONS+MBEDTLS_ERR_ASN1_LENGTH_MISMATCH
```

Nit: missing newline at EOF

File: tests/suites/test_suite_x509parse.function [hunk 2]
```diff
@@ -1573,58 +1573,69 @@ exit:
...
+        if (crt.authority_key_id.keyIdentifier.len > 0)
+        {
+            TEST_EQUAL(crt.authority_key_id.keyIdentifier.tag, MBEDTLS_ASN1_OCTET_STRING);
+            TEST_EQUAL(memcmp(crt.authority_key_id.keyIdentifier.p, keyId->x, keyId->len), 0);
+            TEST_EQUAL(crt.authority_key_id.keyIdentifier.len, keyId->len);
+        }
 
         /* Issuer test */
-        mbedtls_x509_sequence *issuerPtr = &crt.authority_key_id.authorityCertIssuer;
+        if (crt.authority_key_id.authorityCertIssuer.buf.len > 0)
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_x509parse.function:1603]: Use a authorityKeyId_issuer check instead in case if the test does not parse it properly.

**Comment by AndrzejKurek** [comment]
Also, the CI is unhappy so I'm applying the "needs:ci" label too.

File: tests/suites/test_suite_x509parse.function [hunk 3]
```diff
@@ -1573,58 +1573,69 @@ exit:
...
     mbedtls_x509_crt_init(&crt);
 
-    TEST_ASSERT(mbedtls_x509_crt_parse_der(&crt, buf->x, buf->len) == ref_ret);
+    TEST_EQUAL(mbedtls_x509_crt_parse_file(&crt, file), ref_ret);
 
     if (ref_ret == 0) {
         /* KeyId test */
-        TEST_ASSERT(crt.authority_key_id.keyIdentifier.tag == MBEDTLS_ASN1_OCTET_STRING);
-        TEST_ASSERT(crt.authority_key_id.keyIdentifier.len == keyIdLength);
+        if (crt.authority_key_id.keyIdentifier.len > 0)
```
[the remark this answers is not in the document]
Comment (mprse) [reply tests/suites/test_suite_x509parse.function:1595]: If parsing of `keyIdentifier` fails then we will have an error case. But I agree that using expected values for this check is better.

File: tests/suites/test_suite_x509parse.data [hunk 4]
```diff
@@ -3280,37 +3280,61 @@ depends_on:MBEDTLS_X509_CRT_PARSE_C
...
+depends_on:MBEDTLS_MD_CAN_SHA256:MBEDTLS_RSA_C
+x509_crt_parse_authoritykeyid:"data_files/authorityKeyId_subjectKeyId_issuer_tag1_malformed.crt.der":"":"":"":MBEDTLS_ERR_X509_INVALID_EXTENSIONS+MBEDTLS_ERR_ASN1_UNEXPECTED_TAG
 
 X509 CRT parse Authority Key Id - Wrong Issuer Tag 2
```
[the remark this answers is not in the document]
Comment (mprse) [reply tests/suites/test_suite_x509parse.data:3330]: Wrong DirectoryName tag in issuer field ?

File: tests/suites/test_suite_x509parse.function (hunk 3 above)
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply tests/suites/test_suite_x509parse.function:1595]: What I meant is - if there's no field to parse, then there's no error, so we won't it up.

File: tests/suites/test_suite_x509parse.data (hunk 4 above)
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply tests/suites/test_suite_x509parse.data:3330]: Yes :)

File: tests/data_files/Makefile [hunk 5]
```diff
@@ -393,6 +393,48 @@ rsa_pkcs8_2048_public.der: rsa_pkcs8_2048_public.pem
...
+	$(OPENSSL) req -x509 -nodes -days 7300 -key server2.key -outform DER -out $@ -config authorityKeyId_subjectKeyId.conf -extensions 'v3_req'
+
+authorityKeyId_no_keyid.crt.der:
+	$(OPENSSL) req -x509 -nodes -days 7300 -key server2.key -outform DER -out $@ -config authorityKeyId_subjectKeyId.conf -extensions 'v3_req_authorityKeyId_no_keyid'
+
+authorityKeyId_no_authorityKeyId.crt.der:
+	$(OPENSSL) req -x509 -nodes -days 7300 -key server2.key -outform DER -out $@ -config authorityKeyId_subjectKeyId.conf -extensions 'v3_req_no_authorityKeyId'
+
+authorityKeyId_empty.crt.der:
+	$(OPENSSL) req -x509 -nodes -days 7300 -key server2.key -outform DER -out $@ -config authorityKeyId_subjectKeyId.conf -extensions 'v3_req_authorityKeyId_empty'
```
Comment (AndrzejKurek) [inline tests/data_files/Makefile:406]: This is not an empty AuthorityKeyIdentifier, it's missing, as in the case above.

File: tests/data_files/authorityKeyId_subjectKeyId.conf [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+[v3_req]
+subjectKeyIdentifier = hash
+authorityKeyIdentifier = keyid:always,issuer:always
+[v3_req_authorityKeyId_no_keyid]
+subjectKeyIdentifier = hash
+authorityKeyIdentifier = issuer:always
+[v3_req_no_authorityKeyId]
+subjectKeyIdentifier = hash
+[v3_req_authorityKeyId_empty]
+subjectKeyIdentifier = hash
```
Comment (AndrzejKurek) [inline tests/data_files/authorityKeyId_subjectKeyId.conf:18]: This is not an empty AuthorityKeyIdentifier, it's missing, as in the case above.

**Review by AndrzejKurek — DISMISSED** [review]

**Comment by mprse** [comment]
Rebased to resolve conflicts (`makefile`).

**Review by AndrzejKurek — DISMISSED** [review]

File: tests/suites/test_suite_x509parse.function [hunk 7]
```diff
@@ -1549,82 +1549,97 @@ exit:
...
-                result |=
-                    (authorityKeyId_issuer[bufferCounter++] != pname->val.p[issuerCounter]);
+            while (pname != NULL) {
+                for (issuerCounter = 0; issuerCounter < pname->val.len; issuerCounter++) {
+                    result |=
+                        (authorityKeyId_issuer[bufferCounter++] != pname->val.p[issuerCounter]);
+                }
+                bufferCounter++; /* Skipping the slash */
+                pname = pname->next;
             }
```
Comment (davidhorstmann-arm) [inline tests/suites/test_suite_x509parse.function:1620]: Not a blocker for this PR, but I'd prefer using standard functions for comparison of directory names.

I think we can either:
* Use `mbedtls_x509_string_to_names()` to convert the test string to a DN, then use `x509_name_cmp()` to compare to the name in the cert. This requires exporting `x509_name_cmp()` publicly.
* Use `mbedtls_x509_dn_gets()` to convert the name in the cert to a string and `strcmp()` with the test name.

But I'd be happy for this change to be in a followup PR.

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
One issue after regenerating the certificates, one cosmetic non-blocker.

Looks good otherwise!

File: tests/data_files/Makefile [hunk 8]
```diff
@@ -550,6 +550,48 @@ crl_cat_rsa-ec.pem:crl.pem crl-ec-sha256.pem
 
 all_final += crl_cat_ec-rsa.pem crl_cat_rsa-ec.pem
 
+authorityKeyId_subjectKeyId.crt.der:
+	$(OPENSSL) req -x509 -nodes -days 7300 -key server2.key -outform DER -out $@ -config authorityKeyId_subjectKeyId.conf -extensions 'v3_req'
```
[the remark this answers is not in the document]
Comment (mprse) [reply tests/data_files/Makefile]: Fixed as suggested.

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM, thanks!

**Review by AndrzejKurek — APPROVED** [review]

**Pull request merged**